### PR TITLE
[ fix #3874 ] updated core language in user manual

### LINKED ISSUE #3874: Core Syntax Documentation
I am having trouble finding in the documentation an unitary description of the core syntax of Agda. I found these by now:
- the [Core language](https://agda.readthedocs.io/en/latest/language/core-language.html) in readthedocs, self-labeled as stub, where a **Term** is defined in Agda
- an old [Core Syntax](https://wiki.portal.chalmers.se/agda/pmwiki.php?n=ReferenceManual.CoreSyntax) in the wiki about the grammar of terms explained in another way
- the [Syntax Overview](https://wiki.portal.chalmers.se/agda/pmwiki.php?n=DevelopersManual.SyntaxOverview) in the wiki that explains in plain english the three key components of the syntax (expressions, declarations and pragmas) and the three levels of interpretation (concrete, abstract and internal) but in the [Syntax folder](https://github.com/agda/agda/tree/master/src/full/Agda/Syntax) seems to be many more levels of interpretation

Am i missing something? Is this explained somewhere else?

### PR DESCRIPTION
**Pull request by pnlph**
added grammar and the different syntax levels.

### CODE REVIEW

**Comment by pnlph** [comment]
Makefile returns error 193, I guess this is because the code blocks, any ideas on how to fix it?

@andreasabel do you have an updated version of the grammar? I took that one from the wiki.

File: doc/user-manual/language/core-language.lagda.rst [hunk 1]
```diff
@@ -8,21 +8,182 @@
...
+     ==[ nicifier (Syntax.Concrete.Definitions) ]==>
+  'Nice' concrete syntax
+     ==[ scope checking (Syntax.Translation.ConcreteToAbstract) ]==>
+  Abstract syntax
+     ==[ type checking (TypeChecking.Rules.*) ]==>
+  Internal syntax
+     ==[ Agda.Compiler.ToTreeless ]==>
+  Treeless syntax
+     ==[ different backends (Compiler.MAlonzo.*, Compiler.JS.*, ...) ]==>
+  Executable
```
Comment (andreasabel) [inline doc/user-manual/language/core-language.lagda.rst:90]: The backends only produce source code (Haskell, JS, ...).  To create an executable, there might be another stage (e.g. calling the GHC compiler).

File: doc/user-manual/language/core-language.lagda.rst [hunk 2]
```diff
@@ -8,21 +8,182 @@
 Core language
 *************
 
-.. note::
-   This is a stub
+A program in Agda consists of a number of declarations written in an ``*.agda``
+file. A declaration introduces a new identifier and gives its type and
+definition. It is possible to declare:
+
+* :ref:`datatypes <data-types>`
```
Comment (andreasabel) [inline doc/user-manual/language/core-language.lagda.rst:15]: Think on how to order these logically (or alt. alphabetically).

File: doc/user-manual/language/core-language.lagda.rst [hunk 3]
```diff
@@ -8,21 +8,182 @@
...
+
+Grammar
+-------
+
+At its core, Agda is a dependently typed lambda calculus. The grammar of terms
+is as follows:
+
+.. code-block:: text
+
+  a ::= x                       -- variable
```
Comment (andreasabel) [inline doc/user-manual/language/core-language.lagda.rst:44]: This is incomplete.  Internally, we have everything that you show below with the `Term` type.

File: doc/user-manual/language/core-language.lagda.rst [hunk 4]
```diff
@@ -8,21 +8,182 @@
...
+          --   and IApply only for data constructors.
+          | Pi (Dom Type) (Abs Type)      -- ^ dependent or non-dependent function space
+          | Sort Sort
+          | Level Level
+          | MetaV {-# UNPACK #-} !MetaId Elims
+          | DontCare Term
+            -- ^ Irrelevant stuff in relevant position, but created
+            --   in an irrelevant context.  Basically, an internal
+            --   version of the irrelevance axiom @.irrAx : .A -> A@.
+          | Dummy String Elims
```
Comment (andreasabel) [inline doc/user-manual/language/core-language.lagda.rst]: `DontCare` and `Dummy` do not need to be shown in the user manual.

I suppose the internal syntax should be shown as a grammar rather than as Haskell code.

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks.

I wonder whether part of this (like the pipeline) is more relevant to developers.  Maybe a new chapter on "developer documentation" rather than a section under "language" would be a better place?

File: doc/user-manual/language/core-language.lagda.rst (hunk 4 above)
Comment (pnlph) [reply doc/user-manual/language/core-language.lagda.rst]: OK, this was inherited from the old stub. I will update the grammar above.

File: doc/user-manual/language/core-language.lagda.rst (hunk 2 above)
Comment (pnlph) [reply doc/user-manual/language/core-language.lagda.rst:15]: I think logically would be better, is this correct?

* modules
* program options
* postulates
* variables
* datatypes
* record types
* pragmas
* function definitions
* local definitions
* pattern-synonyms
* precedence

Comment (andreasabel) [reply doc/user-manual/language/core-language.lagda.rst:15]: I could imagine going from frequent to exotic.
data, record, function, modules, local, postulates, variables, pattern-synonyms, precedence, pragmas, options

File: doc/user-manual/language/core-language.lagda.rst (hunk 3 above)
Comment (pnlph) [reply doc/user-manual/language/core-language.lagda.rst:44]: I updated the grammar, but I am a bit lost when trying to translate the haskell term into BNF-style rules, can you help?

File: doc/user-manual/language/core-language.lagda.rst (hunk 4 above)
Comment (pnlph) [reply doc/user-manual/language/core-language.lagda.rst]: I rewrote the introductory sentence to the haskell term, I think it would be good if we show that it has the same elements as the grammar, what do you think?

**Comment by pnlph** [comment]
> I wonder whether part of this (like the pipeline) is more relevant to developers.

I changed the formulation, I think this information could also be relevant for the users that want to know more about how the translation works...

> Maybe a new chapter on "developer documentation" rather than a section under "language" would be a better place?

...but it could be splitted to another chapter. By "developer documentation" you mean the section `Contribute -> Documentation`?

File: doc/user-manual/language/core-language.lagda.rst (hunk 3 above)
Comment (andreasabel) [reply doc/user-manual/language/core-language.lagda.rst:44]: We did some initial work on an Agda specification: https://agda.github.io/agda-spec/core-agda.pdf 
Maybe this could be a good source for the documentation you are writing here.

File: doc/user-manual/language/core-language.lagda.rst [hunk 5]
```diff
@@ -36,19 +36,19 @@ in :ref:`mutual recursion <mutual-recursion>`.
...
 
 .. code-block:: text
 
   a ::= x                       -- variable
-      | λ x → a                 -- abstraction
-      | ?                       -- literal
-      | def ?                   -- definition
+      | λ x → a                 -- lambda abstraction
+      | f                       -- defined function
       | a a                     -- application
```
Comment (andreasabel) [inline doc/user-manual/language/core-language.lagda.rst]: Agda internally does not have binary application.

File: doc/user-manual/language/core-language.lagda.rst [hunk 6]
```diff
@@ -36,19 +36,19 @@ in :ref:`mutual recursion <mutual-recursion>`.
...
       | (x : a) → a             -- function space
-      | Set[n]                  -- universe with polymorphism
-      | Setω[n]                 -- universe without polymorphism
+      | F                       -- data/record type
+      | c a                     -- data/record constructor
+      | s                       -- sort Seti, Setω+i
       | (a)                     -- grouping
```
Comment (andreasabel) [inline doc/user-manual/language/core-language.lagda.rst]: Parentheses are not part of an abstract syntax (only concrete syntax).

**Review by andreasabel — CHANGES_REQUESTED** [review]

**Review by JacquesCarette — APPROVED** [review]
This is a very nice step forward in documenting these things, thanks.

File: doc/user-manual/language/core-language.lagda.rst [hunk 7]
```diff
@@ -8,21 +8,177 @@
...
+desugaring at all.  This is what the parser produces. The idea is that if we
+figure out how to keep the concrete syntax around, it can be printed exactly
+as the user wrote it.
+
+Nice Concrete Syntax
+--------------------
+
+The translation from "Agda.Syntax.Concrete" to "Agda.Syntax.Abstract" involves
+scope analysis, figuring out infix operator precedences and tidying up
+definitions.
```
Comment (gallais) [inline doc/user-manual/language/core-language.lagda.rst]: This should probably go into the `Abstract Syntax` part.

The `Nice Concrete Syntax` is a slightly reorganised concrete syntax
that is easier to deal with internally. This is, among other things,
where new-style mutual blocks are detected, where definitions are
assembled from their isolated parts, where fixity information is
collected and attached to definitions, and where a lot of warnings
are emitted for valid but probably unintended things that are essentially
dead code such as empty `instance` blocks and misplaced pragmas.

File: doc/user-manual/language/core-language.lagda.rst [hunk 8]
```diff
@@ -8,21 +8,177 @@
...
+
+The abstract syntax ``Agda.Syntax.Abstract`` is the result after desugaring
+and scope analysis of the concrete syntax. The type checker works on abstract
+syntax, producing internal syntax.
+
+Internal Syntax
+---------------
+
+This is the final stage of syntax before being handed off to one of the
+backends. Terms are well-scoped and well-typed.
```
Comment (gallais) [inline doc/user-manual/language/core-language.lagda.rst:160]: They are also checked for safety (termination and coverage checking
for functions, positivity checking for datatypes).

The type-directed operations (instance resolution and disambiguation
of overloaded constructors for instance) happen here.

File: doc/user-manual/language/core-language.lagda.rst [hunk 9]
```diff
@@ -8,21 +8,177 @@
...
+---------------
+
+The treeless syntax is intended to be used as input for the
+:ref:`compiler backends <compiler-backends>`. It is more low-level than the
+internal syntax and is not used for type checking. Some of the features of
+the treeless syntax are:
+
+* case expressions instead of case trees
+* no instantiated datatypes / constructors
+
```
Comment (gallais) [inline doc/user-manual/language/core-language.lagda.rst:200]: An interesting feature of treeless is that it is in so-called [`A-normal form`](https://en.wikipedia.org/wiki/A-normal_form).

That means that all the case expressions are targeting a *single*
variable and all alternatives may only peel off *one* constructor.
This make it a lot simpler to handle in the backends than a language
where you may case on arbitrary expressions & use deep patterns.

**Review by gallais — APPROVED** [review]
Threw in a couple of things I know about the various stages, their purposes
and features. This PR is a very welcome addition to the documentation!

**Comment by pnlph** [comment]
I think all changes are now updated, feel free to reopen resolved conversations if not.

**Review by andreasabel — APPROVED** [review]

**Comment by andreasabel** [comment]
I am merging this now, we can always fix/improve more later.
